### DISCUSSION TITLE
Hiding a slide

### DISCUSSION THREAD

**Discussion by Juanvvc** [marp-team/marp#162] [Ideas] [answered]
Sometimes, I'd want to hide a slide from a presentation without removing it, for future references or just change slightly a presentation depending on the audience.

I know I could comment the slide using `<!-- -->`, but it would involve commenting in / out all extra slides every time.

Is there any way to hide slides in Marp?

I'm thinking about something like this, i.e., using a custom CSS class:

```markdown
<style>
    .extra-slide { display: none}
</style>

# Normal slide

Text

---

# Hidden slide
<!-- _class: extra-slide -->

Text

---

# Normal slide

Text
```

This class hides the slide... using a black screen when it is the time of showing that slide.

Is there any chance that navigating the presentation using right/left keys would jump over these hidden slides?

Extra point if navigating with (for example) SHIFT+right/left keys do show these extra slides, in case someone has a question involving them

It is OK for me if the pagination jumps from slide number 3 to 5, for example, showing that there is an extra slide there.

**Comment by yhatt** [comment]
We have thought using HTML comment is enough to hide a slide.

```markdown
# Normal slide

Text

<!--
---

# Hidden slide

Text
-->
---

# Normal slide

Text
```

Even so, I think that switching visibillity of the slide for navigation is a common case and valuable (e.g. [Slide visibillity in reveal.js](https://revealjs.com/slide-visibility/)). If received many requests, we will consider to implement in [Marpit framework](https://marpit.marp.app). 

> FYI: Setting `display: none` is just meaning the slide is not visible, and not meaning to skip navigation. (Marp slides can be avilable for any other purpose than presentation on the stage so a style should not make a specific meaning as much as possible).

**Reply by undergroundwires** [reply]
Looking forward to have this in the framework. It's one of the core features that helps with reusable (context-switchable) slides by those who often do presentations.

HTML comments do not work when there are speaker notes or other _class directives and similar in the slide.

**Comment by tsaquet** [comment]
Hello @Juanvvc ! I need to do the same thing, did you find a way to do what you need ? Thank you :)

**Comment by Juanvvc** [comment]
Hello Thomas,

I'm afraid I didn't find a solution, but having the possibility of using
"hidden / extra slides" is a feature I would be very interested in, for the
academia. Extra slides with additional explanations or details is very
useful for the students. Also, I would also appreciate the possibility of
showing these extra slides, during the lecture, if necessary.

I believe the best option is capturing the event en JavaScript. For
example, sending two "next slide" events if the next slide is "hidden",
except if the shift key is clicked.

I've not explored this possibility yet.

Regards

**Comment by tsaquet** [comment]
Hello Juan,

Thank you for your answer. I upvoted the request, I hope it will be added to the roadmap :)

I needed this feature to share groups of slides between decks and being able to hide some slides in the group.

To solve my problem I wrote a golang program that takes a `json` file describing which slides I need in a deck as input and generates an input `.md` file for Marp. For now it will be enough !

**Comment by yhatt** [comment, marked as answer]
Often overlooked but the idea can realize with a relatively simple Marp plugin, as you can see from `better-for-plugin` label of this thread. This is a plugin example to add `hide` local directive support into Marp: https://gist.github.com/yhatt/13a9f52c96821ae34c3de5f732f305df

We have hoped Marp community's someone passionate about the idea is going to make a Marp plugin, publish to npm, and maintain. If so, users who want an extended directive would be able to try it in Marp CLI and Node.js ecosystem easily.

> Marp team could like to keep simple and smart core. We have thought just using HTML comment `<!-- -->` to hide slides is better, also in the view of CommonMark compatibility. So the extended syntax should be a community's plugin until verified its usefulness in widely.
> https://github.com/marp-team/.github/blob/master/CONTRIBUTING.md#before-adding-new-feature

**Reply by tsaquet** [reply]
This works very well, thanks a lot ! Indeed, it would be great if someone could make a maintained plug-in as you suggested. I can't, I don't know the ecosystem (npm, node & co) and reading your code I realize that someone already knowing marp is needed. 
Meanwhile, what you did can be very helpful to people asking for it, I hope they'll find this discussion !

**Reply by StefanoCecere** [reply]
fantastic... why don't you merge it in core Marp?

**Reply by ryanwwest** [reply]
I'd also appreciate if this could be merged into core Marp, or more specifically, if there could be a way to use this plugin or get the hiding slide feature in Marp for VS Code. As while this plugin is a solution, it isn't a solution within VS Code since [Marp for VS Code explicitly doesn't support plugins](https://github.com/orgs/marp-team/discussions/543), and hiding a slide doesn't seem like a highly developer-focused feature.

**Reply by yhatt** [reply]
@StefanoCecere @ryanwwest To integrate that into Marp Core, we need that the effectiveness of new syntax must be accepted to the most of Marp users through the plugin. Check out [our contributing guide](https://github.com/marp-team/.github/blob/master/CONTRIBUTING.md).

> For implementing new syntax and new feature to offical tools, we require able to confirm a worth through a created third-party plugin and many real-world presentations using it. We will consider integrating the plugin as built-in feature if the most of Marp community shows that using the plugin is valuable.
>
> _&mdash; https://github.com/marp-team/.github/blob/master/CONTRIBUTING.md#before-adding-new-feature_

The first step is  to publish the plugin to npm by Marp community, and making it easier for CLI users to test and verify like following:

```javascript
// engine.mjs
import marpPluginHideSlides from 'marp-plugin-hide-slides'

export default ({ marp }) => marp.use(marpPluginHideSlides)
```

```shell
npm install --save-dev @marp-team/marp-cli marp-plugin-hide-slides
npx marp --engine engine.mjs input-markdown.md
```

If use cases for the plugin are sufficiently confirmed and the syntax has no issues, we may be considered for integration into the core.

**Reply by Beurt** [reply]
This plugin is great, but I cannot make it work with a PDF output. Here is the CLI output:

```bash
[  INFO ] Converting 1 markdown...
[  WARN ] Insecure local file accessing is enabled for conversion from test.temp.md.
[ ERROR ] Failed converting Markdown. (`index` must be at least 0 and at most 2, but was actually 3)
```

**Reply by Beurt** [reply]
The problem seems to be with the [`splice`](https://gist.github.com/yhatt/13a9f52c96821ae34c3de5f732f305df#file-hide-slides-plugin-js-L29) but I can't figure out why. The removed tokens seem to be the right ones. And what is strange is that the bug is only for a PDF output, not for an HTML output.

**Reply by Beurt** [reply]
Investigating further.

It came to my mind that I am using marp in a bash script that adds `--pdf-outlines.pages` option when the output is PDF. And this is where the problem is.

I don't know if I am skilled enough to lookup for the bug into the `outline-plugin.ts` but both hiding slide and pdf-outlines plugin are not interacting well.
